### PR TITLE
Execute checkout step inside docker

### PR DESCRIPTION
**Pull request by stsewd**
This is a POC from https://github.com/readthedocs/readthedocs.org/issues/5673#issuecomment-510709478

I'm making a PR so is easy to discuss some problems? Or hope so.

#### How to install/copy the script?

Currently, I'm just mounting a volume inside the container, but we can make it an instalable package.

#### How to install the deps?

Currently, I'm installing gitpython in the _global_ python environment of the container, but we can create a virtualenv just for this package (we need to refactor our class that creates virtualenvs for this).

#### Other problems

We have chicken/egg problem, we don't know the build image to use before the checkout step, so we'll be using two build env, one for the checkout process (we can use any docker image we want) and other for the normal build process. We can keep it this way and initialize the ssh agent on each image or use the providers API to query the config file before cloning. The second option introduce more changes and removes support for repos that are not hosted on those providers, so I think we are just fine with the first option for now (this may be a problem when we move to lambda functions I think?)

### CODE REVIEW

File: rtd_scripts/get_vcs_info.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python
```
Comment (humitos) [inline rtd_scripts/get_vcs_info.py]: This file should be named mentioning `git` on its name.

**Comment by stsewd** [comment]
> How are we going to share the cloned code between these two images?

We have a shared volume, is the same as executing the checkout step as now.

> Another thing to consider here is that with this PR we are only solving the git case, but we need to do the same for the others as well --where we are using some custom code that we wrote for the parsing.

Those commands are executed inside the build environment (docker or local), we only need to do this with git because we are using a library (gitpython)

> Finally, I think that I lost the main goal of all of this. What's the problem we are solving by running git inside the container? I'd like to have it 100% isolated, but it seems that we are adding some issues (and I'm not sure what we are solving)

We want to be able to run the ssh agent inside docker for .com.

**Comment by humitos** [comment]
> Those commands are executed inside the build environment (docker or local), we only need to do this with git because we are using a library (gitpython)

So, considering that we already hit some gitpython bugs and that all this code/PR is needed only for git, isn't it better to go back to our own implementation that parses the git command output and forget about this?

**Comment by stsewd** [comment]
> So, considering that we already hit some gitpython bugs and that all this code/PR is needed only for git, isn't it better to go back to our own implementation that parses the git command output and forget about this?

That was an option https://github.com/readthedocs/readthedocs.org/issues/5673
But I remember that Anthony was for the second option too

**Comment by agjohnson** [comment]
I think we need a call on this one to discuss options and this implementation.

On my side: regarding an option that uses GitPython vs parsing the output, I suppose I'd rather rely on GitPython to do this. Mostly for the reason that we'll have to maintain a connection to git/hg no matter what -- GitPython is at least a stronger choice for this job. I'm partially unsure about the connection mechanism between the application and the container. I do feel like there is maybe something more we could do here. The biggest hurdle is how code is organized, it does feel a little lost in a one-off script now. We could explore making this a separate package if it would be cleaner.

**Comment by agjohnson** [comment]
We didn't come to a conclusion here, I think because we all have reservations about the solutions we've noted so far -- and because we don't really have any better opinions.

This change does address a security issue around exploiting git, which would be good to address eventually.

The other part of this work is to unblock sharing an ssh-agent between git checkout and docker operations. This is perhaps the more immediate change we can introduce, and directly makes the user's experience better.

Two options here are:

* Keep the ssh-agent out of docker, pass in the socket so steps can have access to the agent. ssh-agent on the host system dies after timeout
* Put the ssh-agent in docker, expose the socket to the host system for git checkout steps. ssh-agent lifecycle is directly linked to docker container -- it lives the entire time the container is in operation, it dies as soon as the container is killed.

**Comment by ericholscher** [comment]
I guess I want to start from first-principals a bit -- is there a reason we can't run the code we currently have, just inside Docker? That seems to me to be the most obvious first step, but I don't know why that approach doesn't work.

**Comment by agjohnson** [comment]
It wasn't just unicode branches, it was the actual parsing where we weren't consisently parsing the output correctly, and gitpython does this all for us. It does seem like a lot of duplication of effort for a jankier solution.

We can also be looking into gitpython/hg internals and see if it's possible to replace the execution method, or to parse a different i/o stream that the system i/o (i/o from docker exec in our case).

**Comment by stsewd** [comment]
Closing in favor of https://github.com/readthedocs/readthedocs.org/pull/6436

There is still one decision, related to use the  docker image just with git or use our build images

**Pull request closed without merging**